### PR TITLE
Document destination diagrams

### PR DESCRIPTION
**Pull request by pierrechevalier83**

### CODE REVIEW

File: routing_model/index.html [hunk 1]
```diff
@@ -102,6 +102,15 @@ <h2>Node's entire lifetime</h2>
 
 
   <h2>Node as a member of a section</h2>
+  <p>Once a node has joined a section, they need to be ready to take on multiple roles simultaneously:<br>
+  <ul>
+    <li> Deal with rellocation:<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: typo. "relocation"
Multiple places, same for rellocate

File: routing_model/index.html [hunk 2]
```diff
@@ -102,6 +102,15 @@ <h2>Node's entire lifetime</h2>
...
+         The section as a whole will sometimes act as the source and sometimes as the destination of rellocations. This means nodes will sometimes rellocate away from this section and sometimes rellocate to this section. Nodes in this section will need to perform the right flow to deal with both of these situations.
+    <li> Decide who's an elder or a plain old adult
+  </ul>
+  All of these flows are happening simultaneously, but really, only one flow is active at anytime and all others must be in a "wait" state.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: I am not sure it is clear that for example, a node can handle relocating a node to its section at the same time it is relocating a node away from its section.

File: routing_model/index.html [hunk 3]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
   <h2>Processing a candidate trying to join our section</h2>
+  <p>This flow represents what we do when a candidate contacts us to join our section.<br>
+  The process starts as we receive an ExpectCandidate RPC from this node.<br>
+  We vote for it in PARSEC to be sure all members of the section process it in the same order.<br>
+  Once it reaches consensus, we are ready to process that candidate by letting them perform the resource_proof (see AcceptAsCandidate).<br>
+  There are some subtleties, such as the fact that we only want to process one candidate at a time, but this is the general idea.
+  </p>
+  <button class="collapsible">RPC: ExpectCandidate</button>
+  <div class="content">
+    <p>We receive this RPC from a node that wants to join our section.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: We receive this RPC from a section that want to relocate a node to our section.

File: routing_model/index.html [hunk 4]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+    They aren't connected to us, yet; so the RPC reaches us through their proxy node.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: The node is not communicating with us yet, only once we sent RelocateResponse RPC to the originating section.

File: routing_model/index.html [hunk 5]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+    On receiving it, we vote for ParsecExpectCandidate to process it in the same order as other members of our section.<br>
+    It kickstarts the entire chain of events in this diagram.
```
Comment (jeanphilippeD) [inline routing_model/index.html]: > It kickstarts the entire chain of events in this diagram

...even if only enough node in our section received that event.

File: routing_model/index.html [hunk 6]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+  <div class=content>
+    <p>These RPCs come from AcceptAsCandidate.<br>
+    The same node could be accepted by some nodes who would vote ParsecOnline, but also time out for some other nodes who would vote for ParsecPurgeCandidate.<br>
+    If it's the case, we only want to process the first of these two events and discard the other one.<br>
+    Because the "Wait for 2" loop in AcceptAsCandidate has higher precendence than the "Wait for 1" loop in this one, this is exactly what will happen.
+    </p>
+  </div>
+  <button class="collapsible">shorter_prefix_section</button>
+  <div class="content">
+    <p>If we know of a section that has a shorter prefix than ours, we prefer for them to receive this incoming node rather than ourselves as it will help kepp the Network's sections tree balanced.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: kepp -> keep

File: routing_model/index.html [hunk 7]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+    This shorter_prefix_section is a function that will return None if we are the shortest of any section we know, Some if there is a better candidate.<br>
+    If it is Some, we will relay the RPC to them instead of ourselves.
```
Comment (jeanphilippeD) [inline routing_model/index.html]: unclear, we do not relay RPC to ourself

File: routing_model/index.html [hunk 8]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+    </p>
+  </div>
+  <button class="collapsible">PROCESSING_CANDIDATE</button>
+  <div class="content">
+    <p>We want to allow at most one node at a time to enter our section through rellocation.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: We want to relocate at most one node at a time to our section.

File: routing_model/index.html [hunk 9]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+    The local variable: PROCESSING_CANDIDATE allows us to keep track of the fact someone is already passing the resource proof process, so we can't accept anyone else for now.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: > allows us to keep track of the fact someone

=> indicates if someone

File: routing_model/index.html [hunk 10]
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
...
+    When PROCESSING_CANDIDATE is true and we reach consensus on ParsecExpectCandidate, we simply send a RefuseCandidate RPC to the would-be-incoming-node so they know they must try another section or try again later.
```
Comment (jeanphilippeD) [inline routing_model/index.html]: somewhat inaccurate:
- only refuse if we are the shortest_prefix.
- send it to the originating section, not the node.

> we simply

=> we
> so they know they must try another section or try again later.

=> so they can try another section or try again later

File: routing_model/index.html [hunk 11]
```diff
@@ -193,6 +252,28 @@ <h2>Processing a candidate trying to join our section</h2>
   </div>
 
   <h2>Resource proof from a destination's point of view</h2>
+  <p>In the previous diagram, we ensured an incoming candidate would only be processed for resource_proof if we are the best fit we know and we are not currently processing another node.<br>
+  This leads us here: to the resource proof.<br>
+  We will first add this node to our peer_list with the sate ResourceProof and send them a RelocateResponse.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: RelocateResponse => To source section

File: routing_model/index.html [hunk 12]
```diff
@@ -193,6 +252,28 @@ <h2>Processing a candidate trying to join our section</h2>
...
+  This will let them know we are ready to process them.<br>
+  They will then send their CandidateInfo to everyone in our section.<br>
+  On receiving a valid CandidateInfo, I will send them a ResourceProof RPC. This gives them the "problem to solve". As they solve it, they will send me ResourceProofResponses. These will be parts of the proof. On receiving valid parts, I must send a ResourceProofReceipt. Once they finally send me the last valid part, they passed their resource proof and I vote for ParsecOnline (essentially accepting them as a member of my section).<br>
+  At any time during this process, they may timeout, in which case I will decided to reject them and vote for ParsecPurgeCandidate.<br>
+  This process ends once I reach consensus on either accepting or the candidate (ParsecOnline) or refusing them (ParsecPurgeCandidate).<br>
+  It is possible that both reach consensus, in which case the second one will be discarded in the flow above.
+  </p>
```
Comment (jeanphilippeD) [inline routing_model/index.html:265]: It is possible for a node to exit this process having done nothing more than seen ParsecOnline or ParsecPurgeCandidate.

**Review by jeanphilippeD — CHANGES_REQUESTED** [review]
Looks very good.
There were a couple of inaccuracy, so I'll wait to merge it.

File: routing_model/index.html (hunk 2 above)
Comment (pierrechevalier83) [reply routing_model/index.html]: Uhm. I wasn't 100% clear on this, but I thought that's what you meant. So what about:
> All of these flows are happening simultaneously, but in most cases, only one flow is active at anytime while the others are in a "wait" state.

Or would you just remove the entire sentence?

File: routing_model/index.html (hunk 8 above)
Comment (pierrechevalier83) [reply routing_model/index.html]: Uhm. I was trying to emphasize the "incoming" aspect... But I guess yours reads better...
Maybe some middle-ground:
> We want to accept only one incoming relocation at a time into our section.

File: routing_model/index.html (hunk 12 above)
Comment (pierrechevalier83) [reply routing_model/index.html:265]: Well, consensus has to be reached, hasn't it?
Are you trying to say that these have ConsensusMode::Single, so a supermajority is not needed, which is why both may reach consensus?

**Comment by pierrechevalier83** [comment]
Thanks for the detailed review JP. I've addressed your comments. For a couple of them, I didn't exactly follow your recommendation and explained why in a comment. I'll wait for your confirmation that you're fine with it now :)

File: routing_model/index.html (hunk 12 above)
Comment (jeanphilippeD) [reply routing_model/index.html:265]: no, I'm trying to say that consensus can be reached without any input for that node.

File: routing_model/index.html (hunk 2 above)
Comment (jeanphilippeD) [reply routing_model/index.html]: > All of these flows are happening simultaneously, but in most cases, only one flow is active at anytime while the others are in a "wait" state.

No, in all case, only one flow is active at a time, but there are multiple events to see and wait for in a flow, so you may have sent a RelocateResponse and are resource proofing, but in the wait, you may see the relocation trigger and send yourself an ExpectCandidate.


I'm not sure I would remove the sentence, the original sentence was not wrong.

Comment (jeanphilippeD) [reply routing_model/index.html]: I'm happy with the sentence as it is.

File: routing_model/index.html [hunk 13]
```diff
@@ -132,8 +141,59 @@ <h2>Node as a member of a section</h2>
 </div>
 
   <h1>Destination section</h1>
-  Receiving a node that is being relocated.
+  <p>As a member of a section, our section will sometimes receive a node that is being relocated. These  diagrams are from the point of view of one of the nodes in the section, doing its part to handle the node that is trying to relocate to this section.
+  </p>
   <h2>Processing a candidate trying to join our section</h2>
+  <p>This flow represents what we do when a candidate contacts us to join our section.<br>
+  The process starts as we receive an ExpectCandidate RPC from this node.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html:148]: > This flow represents what we do when a candidate contacts us to join our section.
> The process starts as we receive an ExpectCandidate RPC from this node.

The flow is from a section contacting us with a candidate. ExpectCandidate is from the section.

File: routing_model/index.html [hunk 14]
```diff
@@ -132,8 +141,59 @@ <h2>Node as a member of a section</h2>
...
+  <button class="collapsible">RPC: ResourceProofResponse/<br>CandidateInfo</button>
+  <div class="content">
+  <p>These RPCs should generally be handled during resource proof, in AcceptAsCandidate.<br>
+  It could be that they reach us before we are ready to receive them. For instance, because although enough other nodes got consensus on ParsecExpectCandidate, I'm still a bit behind.<br>
+  It's OK to discard the RPC in this case as the node who aims to be relocated here will perceive this as a timeout from our end and resend later.
+  </p>
+  </div>
+  <button class="collapsible">ParsecOnline<br>ParsecPurgeCandidate</button>
+  <div class=content>
+    <p>These RPCs come from AcceptAsCandidate.<br>
```
Comment (jeanphilippeD) [inline routing_model/index.html]: > These RPCs come from AcceptAsCandidate.

These consensused events come from AcceptAsCandidate

**Review by jeanphilippeD — APPROVED** [review]

**Pull request merged**